Index: kde-runtime/kurifilter-plugins/shorturi/kshorturifilter.cpp
===================================================================
--- kde-runtime.orig/kurifilter-plugins/shorturi/kshorturifilter.cpp	2014-01-21 17:18:45.895102067 -0300
+++ kde-runtime/kurifilter-plugins/shorturi/kshorturifilter.cpp	2014-01-22 14:00:20.347354650 -0300
@@ -279,10 +279,10 @@
     // Environment variable expansion.
     if ( sEnvVarExp.indexIn( path ) == 0 )
     {
-      const char* exp = qgetenv( path.mid( 1, sEnvVarExp.matchedLength() - 1 ).toLocal8Bit().data() );
-      if(exp)
+      QByteArray exp = qgetenv( path.mid( 1, sEnvVarExp.matchedLength() - 1 ).toLocal8Bit().data() );
+      if(! exp.isNull())
       {
-        path.replace( 0, sEnvVarExp.matchedLength(), QString::fromLocal8Bit(exp) );
+        path.replace( 0, sEnvVarExp.matchedLength(), QString::fromLocal8Bit(exp.constData()) );
         expanded = true;
       }
     }
