Index: kde-runtime/knetattach/knetattach.desktop
===================================================================
--- kde-runtime.orig/knetattach/knetattach.desktop	2015-02-23 22:57:53.683998274 +0100
+++ kde-runtime/knetattach/knetattach.desktop	2015-02-23 22:57:53.679998329 +0100
@@ -186,3 +186,4 @@
 Categories=Qt;KDE;Network;RemoteAccess;
 X-DocPath=knetattach/index.html
 OnlyShowIn=KDE;
+NoDisplay=true
